snuba/clickhouse/translators/snuba/mapping.py

```diff
@@ -0,0 +1,152 @@
+
+    This is because only some pairs of expression types (Snuba,
+    Clickhouse) are allowed during translation so we can guarantee any
+    configuration provided by the user will either produce a valid AST
+    or refuse to translate.
+
+    See allowed.py for the valid translation rules and their reasoning.
+    """
+
+    literals: Sequence[LiteralMapper] = field(default_factory=list)
+    columns: Sequence[ColumnMapper] = field(default_factory=list)
+    subscriptables: Sequence[SubscriptableReferenceMapper] = field(default_factory=list)
+    functions: Sequence[FunctionCallMapper] = field(default_factory=list)
+    curried_functions: Sequence[CurriedFunctionCallMapper] = field(default_factory=list)
+    arguments: Sequence[ArgumentMapper] = field(default_factory=list)
+    lambdas: Sequence[LambdaMapper] = field(default_factory=list)
+
+    def concat(self, spec: TranslationMappers) -> TranslationMappers:
+        return TranslationMappers(
+            literals=[*self.literals, *spec.literals],
+            columns=[*self.columns, *spec.columns],
+            subscriptables=[*self.subscriptables, *spec.subscriptables],
+            functions=[*self.functions, *spec.functions],
+            curried_functions=[*self.curried_functions, *spec.curried_functions],
+            arguments=[*self.arguments, *spec.arguments],
+            lambdas=[*self.lambdas, *spec.lambdas],
+        )
+
+
+class SnubaClickhouseMappingTranslator(SnubaClickhouseStrictTranslator):
+    """
+    Translates a Snuba expression into an clickhouse query expression
+    according to a specification provide by the caller.
+
+    The translation of every node in the expression is performed by a
+    series of rules that extend ExpressionMapper.
+    Rules are applied in sequence. Given an expression, the first valid
+    rule for such expression is applied and the result is returned. If
+    no rule can translate such expression an exception is raised.
+    A rule can delegate the translation of its children back to this
+    translator.
+
+    Each rule only has context around the expression provided and its
+    children. It does not have general context around the query or around
+    the expression's ancestors in the AST.
+    This approach implies that, while rules are specific to the
+    relationship between dataset (later entities) and storage, this class
+    keeps the responsibility of orchestrating the translation process.
+
+    It is possible to compose different, independently defined, sets of
+    rules that are applied in a single pass over the AST.
+    This allows us to support joins which can be translated by simply
+    concatenating rule sets associated with each storage involved.
+
+    This relies on a visitor so that we can statically enforce that no
+    expression subtype is added to the code base without properly support
+    it in all translators. The downside is verbosity.
+    """
```

Comment:
I think it may be useful to give some context on how I envision joins to work since it may not be obvious:
phase 1: pre-entity. Here not all columns in the AST will be fully qualified.

non join query: Here the columns now are not fully qualified. It works as it is, columns do not reference a table alias and the rules do not care about table aliases
join query: Here the columns in the ast are fully qualified so that they are not ambiguous. Since the translation rules do not consider table aliases we need to do soemthing about it. The Query Plan Builder for joins will wrap all the translation rules before applying them to account for the table alias. Something like:
```
storage1_rules = [JoinWrapper(rule, table_alias="event") for rule in storage1_rules]
storage2_rules = [JoinWrapper(rule, table_alias="groups") for rule in storage2_rules]
# JoinWrapper takes in the column to translate, removes the alias, execute the inner rule
# and transform the output by adding the alias back.

rules = [*storage1_rules, *storage2_rules]
translator = Translator(rules)
```

phase 2: entities or after #966 
Now we don't care about all that complexity anymore since all columns in the AST (and, as a consequences all translation rules) are fully qualified with the entity name, which replaces the table aliases in joins.

Reply:
This is clever.